[Driver][SYCL] Add initial SYCL offload compilation support

mdtoguchi opened the pull request:
Introduces the SYCL based toolchain and initial toolchain construction when using the '-fsycl' option. This option will enable SYCL based offloading, creating a SPIR-V based IR file packaged into the compiled host object.

This includes early support for creating the host/device object using the new offloading model. The device object is created using the spir64-unknown-unknown target triple.

New/Updated Options:
 -fsycl  Enables SYCL offloading for host and device
 -fsycl-device-only
         Enables device only compilation for SYCL
 -fsycl-host-only
         Enables host only compilation for SYCL

RFC Reference:
https://discourse.llvm.org/t/rfc-sycl-driver-enhancements/74092

Was reverted due to buildbot issues.  Contains additional fixes to pull in the SYCL header dependencies to other toolchains.

This is a reland of: https://github.com/llvm/llvm-project/pull/107493

mdtoguchi committed:
Reland - [Driver][SYCL] Add initial SYCL offload compilation support (#107493)

Introduces the SYCL based toolchain and initial toolchain construction
when using the '-fsycl' option. This option will enable SYCL based
offloading, creating a SPIR-V based IR file packaged into the compiled
host object.

This includes early support for creating the host/device object using
the new offloading model. The device object is created using the
spir64-unknown-unknown target triple.

New/Updated Options:
 -fsycl  Enables SYCL offloading for host and device
 -fsycl-device-only
         Enables device only compilation for SYCL
 -fsycl-host-only
         Enables host only compilation for SYCL

RFC Reference:
https://discourse.llvm.org/t/rfc-sycl-driver-enhancements/74092

Was reverted due to buildbot issues.  Contains additional fixes to pull
in the SYCL header dependencies to other toolchains.

tahonermann reviewed (commented):
This is looking good to me. I added some questions and noted a couple of potential concerns for which I'm not well equipped to have an opinion on. Perhaps others can offer their thoughts on those.

mdtoguchi commented on clang/include/clang/Driver/Options.td:
```diff
@@ -6782,16 +6783,21 @@ defm : FlangIgnoredDiagOpt<"frontend-loop-interchange">;
+  HelpText<"Perform link through clang-sycl-linker via the target "
   "offloading toolchain.">;
+} // let Group = sycl_Group
+
```
I think this can be left to future work.

mdtoguchi commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -6591,6 +6681,18 @@ const ToolChain &Driver::getOffloadingDeviceToolChain(
+        TC = std::make_unique<toolchains::SYCLToolChain>(*this, Target, HostTC,
+                                                         Args);
+        break;
+      default:
+        break;
+      }
     default:
       break;
```
I followed the `OFK_HIP` usage here and added an assert if `TC` isn't populated. https://github.com/llvm/llvm-project/blob/3fdc74687dbf6da831a1c27995d9d7fef3a2a51f/clang/lib/Driver/Driver.cpp#L1063

tahonermann replied:
The `assert()` at that location happens too late. C++ doesn't acknowledge the possibility of a null reference and attempting to construct one results in undefined behavior. A conforming compiler is allowed to assume that taking the address of a reference results in a non-null pointer and may thus elide any actual check for null as an always-false condition; a compiler can therefore conclude that `HIPTC` is always non-null in `assert(HIPTC && ...)` without actually checking. The fix that is actually needed is to `assert(TC && "Could not create offloading device tool chain.")` below before the return of `*TC`. The `assert(HIPTC && ...)` and added `assert(SYCLTC && ...)` statements can then be removed.

mdtoguchi replied:
Thanks for that, I have made updates here: https://github.com/llvm/llvm-project/pull/117268/commits/df928e839dc3350419066ec2b85a365cdf1d9502

tahonermann replied:
Thank you, looks good!

mdtoguchi commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,179 @@
+  // Add the SYCL header search locations in the specified order.
+  //   ../include/sycl/stl_wrappers
+  //   ../include
+  SmallString<128> IncludePath(D.Dir);
+  llvm::sys::path::append(IncludePath, "..");
+  llvm::sys::path::append(IncludePath, "include");
```
I would guess this depends on how ultimately the `include/sycl` headers are installed and how we should point to them.  As the headers are not currently part of the build, this is basically just  a location in which to try and find the SYCL specific headers (this is what we are doing in intel/llvm).

tahonermann replied:
I think we should work this out with the SYCL library folks before making this change then since duplicate paths in the header search path can be problematic and, when they are, difficult to work around.

mdtoguchi replied:
Updates made to remove the header inclusions - `FIXME` in it's place until this is resolved with the library folks.

mdtoguchi commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,179 @@
+  // Diagnose unsupported options only once.
+  for (OptSpecifier Opt : getUnsupportedOpts()) {
+    if (const Arg *A = Args.getLastArg(Opt)) {
+      // All sanitizer options are not currently supported, except
+      // AddressSanitizer.
```
Right - I will remove this special casing until the support is upstreamed.

mdtoguchi committed:
Remove sanitizer support checks

mdtoguchi commented:
Ping

mdtoguchi committed:
Do not add header search directories at this time

mdtoguchi committed:
Adjust offloading toolchain creation check.

tahonermann reviewed (approved):
Thanks, @mdtoguchi, this looks good to me!

tahonermann commented:
@MaskRay, could you please take a look at this PR and give it your stamp of approval if you have no concerns?

mdtoguchi commented:
Ping

MaskRay commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -780,6 +781,41 @@ Driver::OpenMPRuntimeKind Driver::getOpenMPRuntime(const ArgList &Args) const {
+static llvm::Triple getSYCLDeviceTriple(StringRef TargetArch) {
+  SmallVector<StringRef, 5> SYCLAlias = {"spir", "spir64", "spirv", "spirv32",
+                                         "spirv64"};
+  if (std::find(SYCLAlias.begin(), SYCLAlias.end(), TargetArch) !=
```
`llvm::is_contained({..., ...}, TargetArch)`

MaskRay commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -780,6 +781,41 @@ Driver::OpenMPRuntimeKind Driver::getOpenMPRuntime(const ArgList &Args) const {
   return RT;
 }
 
+static const char *getDefaultSYCLArch(Compilation &C) {
```
Remove this trivial function and inline its use

MaskRay commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -993,6 +1028,41 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
     return;
   }
 
+  //
```
We don't use start a comment block with `//\n`. Just delete the part before `We need to generate a SYCL toolchain if the user specified -fsycl.`

MaskRay commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -6591,11 +6679,23 @@ const ToolChain &Driver::getOffloadingDeviceToolChain(
       break;
     }
+    case Action::OFK_SYCL:
+      switch (Target.getArch()) {
```
Use `isSPIR` and `isSPIRV`

MaskRay commented on clang/lib/Driver/ToolChains/Gnu.h:
```diff
@@ -336,6 +338,9 @@ class LLVM_LIBRARY_VISIBILITY Generic_GCC : public ToolChain {
       const llvm::opt::ArgList &DriverArgs,
       llvm::opt::ArgStringList &CC1Args) const override;
 
+  void AddSYCLIncludeArgs(const llvm::opt::ArgList &DriverArgs,
```
since this is new function. name it `add...` instead of `Add...`.

Yes, clangDriver naming is very inconsistent, but we try to make them consistent for new functions.

MaskRay commented on clang/test/Driver/sycl-offload-jit.cpp:
```diff
@@ -0,0 +1,56 @@
+// RUN:   %s 2>&1 \
+// RUN: | FileCheck -check-prefixes=CHK-PHASES %s
+// CHK-PHASES: 0: input, "[[INPUT:.+\.cpp]]", c++, (host-sycl)
+// CHK-PHASES: 1: preprocessor, {0}, c++-cpp-output, (host-sycl)
```
Use `CHK-PHASES-NEXT` whenever applicable

MaskRay commented on clang/test/Driver/sycl-offload-jit.cpp:
```diff
@@ -0,0 +1,56 @@
+
+/// Check the phases graph with -fsycl. Use of -fsycl enables offload
+// RUN: %clang -ccc-print-phases --target=x86_64-unknown-linux-gnu \
+// RUN:   -fsycl %s 2>&1 \
```
unnecessary wrapping here

MaskRay commented on clang/test/Driver/sycl-offload-jit.cpp:
```diff
@@ -0,0 +1,56 @@
+/// well as clang-offload-packager inputs.
+// RUN: %clang -### -fsycl -c --target=x86_64-unknown-linux-gnu %s 2>&1 \
+// RUN: | FileCheck -check-prefix=CHK-DEVICE-TRIPLE %s
+// CHK-DEVICE-TRIPLE: clang{{.*}} "-triple" "spirv64-unknown-unknown"
```
When testing cc1 commands, we sometimes test `"-cc1"`, but we don't test `clang{{.*}}`, since the string could be different in a bazel build.　https://github.com/maskray/config/wiki/LLVM#clanglibdriver

MaskRay commented on clang/test/Driver/sycl-offload-jit.cpp:
```diff
@@ -0,0 +1,56 @@
+/// Check -fsycl-is-device is passed when compiling for the device.
+/// Check -fsycl-is-host is passed when compiling for host.
+// RUN: %clang -### -fsycl -c %s 2>&1 \
+// RUN: | FileCheck -check-prefixes=CHK-FSYCL-IS-DEVICE,CHK-FSYCL-IS-HOST %s
```
indent continuation lines by 2. please fix this everywhere in this PR.

MaskRay commented:
> …(https://github.com/llvm/llvm-project/pull/107493)

The subject line is now wrapped by github's pr system.

 Perhaps just use `[Driver][SYCL] Add initial SYCL offload compilation support` as the subject and mention that this is a reland of #107493.

mdtoguchi committed:
Address a few review comments

 - remove simple function and inline usage
 - use if statement for type
 - use is_contained

mdtoguchi committed:
Update function name to use lowercase first letter

mdtoguchi committed:
Update test to address review comments

mdtoguchi committed:
Merge remote-tracking branch 'origin' into sycl-upstream-driver-reland

mdtoguchi commented:
@MaskRay - thank you for the review.  I have made the updates as suggested, please have another look when time permits.

MaskRay commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,154 @@
+}
+
+// Unsupported options for SYCL device compilation.
+static std::vector<OptSpecifier> getUnsupportedOpts() {
```
return `ArrayRef`. 

UnsupportedOpts below can be a plain array

MaskRay commented on clang/test/Driver/sycl-offload-jit.cpp:
```diff
@@ -0,0 +1,52 @@
+///
```
Thanks for adding a file-level comment. We just write `/// xxxx` instead of `///\n/// xxxx\n///`

MaskRay commented on clang/lib/Driver/ToolChains/MinGW.cpp:
```diff
@@ -630,6 +630,11 @@ void toolchains::MinGW::AddHIPIncludeArgs(const ArgList &DriverArgs,
 
+void toolchains::MinGW::addSYCLIncludeArgs(const ArgList &DriverArgs,
+                                           ArgStringList &CC1Args) const {
+  SYCLInstallation->addSYCLIncludeArgs(DriverArgs, CC1Args);
```
is this needed or tested?

If you add an overload to a specific ToolChain, ensure that this ToolChain gets coverage. Otherwise someone could find the function unneeded and delete it.

mdtoguchi replied:
Initial testing found the need for this due to the general check for the updated header search locations.  As reviews progressed with @tahonermann, it was suggested to not add the explicit header search locations as it was not determined yet (SYCL specific headers and libraries not yet part of the build) so now there is nothing being tested here for this target.  This can be cleaned up and added again later when the headers are actually searched.

mdtoguchi committed:
Clean up comment

mdtoguchi committed:
Use ArrayRef

mdtoguchi committed:
Do not setup additional include steps for toolchain

shiltian commented:
This PR introduced a new compile warning when compiling using clang. Please fix it.

```
llvm-project/clang/lib/Driver/ToolChains/SYCL.cpp:35:10: warning: returning address of local temporary object [-Wreturn-stack-address]
   35 |   return {
      |          ^
   36 |       options::OPT_fsanitize_EQ,      // -fsanitize
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   37 |       options::OPT_fcf_protection_EQ, // -fcf-protection
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   38 |       options::OPT_fprofile_generate,
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   39 |       options::OPT_fprofile_generate_EQ,
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   40 |       options::OPT_fno_profile_generate, // -f[no-]profile-generate
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   41 |       options::OPT_ftest_coverage,
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   42 |       options::OPT_fno_test_coverage, // -f[no-]test-coverage
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   43 |       options::OPT_fcoverage_mapping,
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   44 |       options::OPT_fno_coverage_mapping, // -f[no-]coverage-mapping
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   45 |       options::OPT_coverage,             // --coverage
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   46 |       options::OPT_fprofile_instr_generate,
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   47 |       options::OPT_fprofile_instr_generate_EQ,
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   48 |       options::OPT_fno_profile_instr_generate, // -f[no-]profile-instr-generate
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   49 |       options::OPT_fprofile_arcs,
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~
   50 |       options::OPT_fno_profile_arcs, // -f[no-]profile-arcs
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from llvm-project/clang/lib/Driver/ToolChains/SYCL.cpp:8:
llvm-project/clang/lib/Driver/ToolChains/SYCL.h:27:17: warning: private field 'D' is not used [-Wunused-private-field]
   27 |   const Driver &D;
      |                 ^
2 warnings generated.
```

ilovepi commented:
We're seeing SYCL tests fail on Linux builders after this patch https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-host-linux-x64/b8726474372098645345/overview

```
Exit Code: 1

Command Output (stderr):
--
RUN: at line 1: /b/s/w/ir/x/w/llvm_build/bin/clang -### -fsycl -c /b/s/w/ir/x/w/llvm-llvm-project/clang/test/Driver/sycl.c 2>&1 | /b/s/w/ir/x/w/llvm_build/bin/FileCheck /b/s/w/ir/x/w/llvm-llvm-project/clang/test/Driver/sycl.c --check-prefix=ENABLED
+ /b/s/w/ir/x/w/llvm_build/bin/clang -### -fsycl -c /b/s/w/ir/x/w/llvm-llvm-project/clang/test/Driver/sycl.c
+ /b/s/w/ir/x/w/llvm_build/bin/FileCheck /b/s/w/ir/x/w/llvm-llvm-project/clang/test/Driver/sycl.c --check-prefix=ENABLED
/b/s/w/ir/x/w/llvm-llvm-project/clang/test/Driver/sycl.c:18:13: error: ENABLED: expected string not found in input
// ENABLED: "-cc1"{{.*}} "-fsycl-is-device"
            ^
<stdin>:1:1: note: scanning from here
Fuchsia clang version 20.0.0git (https://llvm.googlesource.com/llvm-project d00f65c6acd9f0e1ddae83391f55eb9d232d2f9e)
^
<stdin>:2:14: note: possible intended match here
Target: x86_64-unknown-linux-gnu
             ^

Input file: <stdin>
Check file: /b/s/w/ir/x/w/llvm-llvm-project/clang/test/Driver/sycl.c

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: Fuchsia clang version 20.0.0git (https://llvm.googlesource.com/llvm-project d00f65c6acd9f0e1ddae83391f55eb9d232d2f9e) 
check:18'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            2: Target: x86_64-unknown-linux-gnu 
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:18'1                  ?                    possible intended match
            3: Thread model: posix 
check:18'0     ~~~~~~~~~~~~~~~~~~~~
            4: InstalledDir: /b/s/w/ir/x/w/llvm_build/bin 
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>
```

Can you take a look, and revert if the fix will take a while?

mdtoguchi commented:
I have posted a fix to address the sanitizer builds and build warning:  https://github.com/llvm/llvm-project/pull/121822

jroelofs commented:
I see a new warning from this patch:
```
clang/lib/Driver/ToolChains/SYCL.cpp:35:10: warning: returning address of local temporary object [-Wreturn-stack-address]
```

jroelofs commented:
also:
```
clang/lib/Driver/ToolChains/SYCL.h:27:17: warning: private field 'D' is not used [-Wunused-private-field]
```
